text/0000-app-mixins.md

```diff
@@ -0,0 +1,130 @@
+* If the mixin name does not have a prefix it will be dynamically added to both the build and run stack images.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When a list of mixins are required by buildpacks via the build plan and the build phase starts:
+
+1. The lifecycle will compare the list of mixins to those provided by the stack using the `io.buildpacks.stack.mixins` label.
+1. If all mixin names are provided by the stack, no further action is required.
+1. If any requested mixin is not provided by the stack, the lifecycle will run the detect phase for all stackpacks defined in the builder.
+1. The lifecycle will compare the mixins added to the build plan to see if they match they required mixins.
+1. If at least one stackpack passes detect and provides the required mixin(s), the lifecycle will execute the stackpack build phase for all passing stackpack(s). If no stackpacks pass detect, or no stackpacks provide the required mixin, the build will fail with a message that describes the mixins that could not be provided.
+1. During the lifecycle's build phase, the stackpacks that passed detection will run against the build and run images accordingly (see details below). All stackpacks will be run before the user's buildpacks.
+
+## Stack Buildpacks
+
+The complete details of a Stack buildpack are described in [another RFC proposal](https://github.com/buildpacks/rfcs/pull/111).
+
+## Rebasing an App
+
+Before a launch image is rebased, the platform must re-run the any stackpacks that were used to build the launch image against the new run-image. Then, the rebase operation can be performed as normal, while including the stackpack layers as part of the stack. This will be made possible by including the stackpack in the run-image, but because the stackpack detect phase is not run, the operation does not need access to the application source.
+
+## Example: ffmpeg
+
+The `project.toml` for an app that requires the `ffmpeg` package might look like this:
+
+```toml
+[project]
+id = "example/image-app"
+name = "My Image App"
+version = "1.0.0"
+
+[build]
+mixins = [ "ffmpeg" ]
+```
+
+## Example: libpq
+
+The `project.toml` for an app that requires the `libpq` package might look like this:
+
+```toml
+[project]
+id = "example/database-app"
+name = "My Database App"
+version = "1.0.0"
+
+[build]
+mixins = [ "build:libpq-dev", "run:libpq" ]
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Installing a mixin a build-time means that the `rebase` must also update the provided mixin. In this way, `rebase` becomes an operation that _may_ do more than edit JSON on a registry. It must also re-run a stack buildpack.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- [Add root buildpack as interface for app image extension RFC](https://github.com/buildpacks/rfcs/pull/77)
+- [App Image Extensions (OS packages)](https://github.com/buildpacks/rfcs/pull/23)
```

Comment:
Would you mind clarifying why these aren't valid alternatives?

Reply:
i wouldn't say they aren't valid approaches. We just don't have support for them.